### PR TITLE
NanoService Reload

### PR DESCRIPTION
**Pull request by smirnovaka**
Not all NanoServices can be recreated on nSpec changes.  For those that cannot, 'reload' has been added to the NanoService interface.  Called from Boot and NSpecFactory to instruct the service that its nSpec was modified and to take appropriate action.

### CODE REVIEW

File: src/foam/nanos/NanoService.js [hunk 1]
```diff
@@ -14,6 +14,18 @@ foam.INTERFACE({
...
+      name: 'reload',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        }
+      ],
+      javaCode: `
+        //add implementation
+      `
```
Comment (lchanmann) [inline src/foam/nanos/NanoService.js]: NanoService is an interface, did you mean to have default implementation for the reload() method?

Comment (jlhughes) [reply src/foam/nanos/NanoService.js]: Yes, as I thought our interfaces supported this now.

Comment (smirnovaka) [reply src/foam/nanos/NanoService.js]: yes, default method implementation on the interface(?). What is wrong with this

Comment (lchanmann) [reply src/foam/nanos/NanoService.js]: It failed to build but on the second look seems like you forget to pass x to the reload call.

Comment (smirnovaka) [reply src/foam/nanos/NanoService.js]: yes, my bad forgot the context. But we do support the default interface implementation

File: src/foam/nanos/boot/NSpecFactory.java [hunk 2]
```diff
@@ -111,49 +131,58 @@ public Object create(X x) {
...
+    logger.warning("Invalidating Service", spec_.getName());
     if ( ! SafetyUtil.equals(spec.getService(), spec_.getService())
       || ! SafetyUtil.equals(spec.getServiceClass(), spec_.getServiceClass())
       || ! SafetyUtil.equals(spec.getServiceScript(), spec_.getServiceScript())
     ) {
-      logger.info("Invalidated Service", spec_.getName());
-      if ( ns_ instanceof DAO ) {
-        // Clustered MDAOs are not reloadable as replay is handled by medusa.
+      if ( ns_ instanceof NanoService ) {
+        Object ns = buildService(x_);
```
Comment (xuerongNanopay) [inline src/foam/nanos/boot/NSpecFactory.java:163]: miss `spec_ = spec;` before calling buildService method.

File: src/foam/nanos/boot/NSpecFactory.java [hunk 3]
```diff
@@ -56,26 +56,49 @@ void buildService(X x) {
...
+    Logger logger = null;
+    if ( ! "logger".equals(spec_.getName()) && ! "PM".equals(spec_.getName()) ) {
+      logger = (Logger) x.get("logger");
+    }
+    if ( logger == null ) {
+      logger = StdoutLogger.instance();
+    }
 
-      Object ns = ns_;
+    X  nx = x_ instanceof SubX ? x_ : x_.getX();
```
Comment (kgrgreer) [inline src/foam/nanos/boot/NSpecFactory.java:100]: Please document what this line is for.

**Comment by kgrgreer** [comment]
I would like to discuss.

**Pull request merged**